src/Workspaces/Core/Portable/FindSymbols/FindReferences/Finders/PreprocessingSymbolReferenceFinder.cs

```diff
@@ -0,0 +1,104 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.LanguageService;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+
+namespace Microsoft.CodeAnalysis.FindSymbols.Finders;
+
+internal class PreprocessingSymbolReferenceFinder : AbstractReferenceFinder<IPreprocessingSymbol>
+{
+    protected override bool CanFind(IPreprocessingSymbol symbol) => true;
+
+    protected sealed override async ValueTask<ImmutableArray<FinderLocation>> FindReferencesInDocumentAsync(
+        IPreprocessingSymbol symbol,
+        FindReferencesDocumentState state,
+        FindReferencesSearchOptions options,
+        CancellationToken cancellationToken)
+    {
+        var tokens = await FindMatchingIdentifierTokensAsync(state, symbol.Name, cancellationToken).ConfigureAwait(false);
+
+        var normalReferences = await FindPreprocessingReferencesInTokensAsync(
+            symbol, state,
+            tokens,
+            cancellationToken).ConfigureAwait(false);
+
+        return normalReferences;
+    }
+
+    protected override async Task<ImmutableArray<Document>> DetermineDocumentsToSearchAsync(
+        IPreprocessingSymbol symbol,
+        HashSet<string>? globalAliases,
+        Project project,
+        IImmutableSet<Document>? documents,
+        FindReferencesSearchOptions options,
+        CancellationToken cancellationToken)
+    {
+        using var _ = ArrayBuilder<Document>.GetInstance(out var resultDocuments);
+
+        // NOTE: We intentionally search for all documents in the entire solution. This is because
+        //       the symbols are validly bound by their requested name, despite their current definition
+        //       state. Therefore, the same symbol name could be shared across multiple projects and
+        //       configured in the project configuration with the same shared identifier.
+
+        var solution = project.Solution;
+        foreach (var solutionProject in solution.Projects)
+        {
+            foreach (var document in await solutionProject.GetAllRegularAndSourceGeneratedDocumentsAsync(cancellationToken).ConfigureAwait(false))
```

Comment:
filter this against 'documents' if present.

Reply:
We intentionally search all the documents in the solution, could `documents` always guarantee this behavior? Do we want to abide to the specific set of `documents` if present, without overriding it?

Reply:
`documents` being null means `search everything`.  `documents being nonnull` means "the callers requested a subset of documents. 

We don't generally make an array with all the docs in it (you are an exception here) because that could literally be hundreds of thousands of items in it :)